### PR TITLE
Added Generalized PoW for CPU

### PR DESCRIPTION
**Pull request by paulhandy**

### CODE REVIEW

File: curl-cpu/src/search.rs [hunk 1]
```diff
@@ -0,0 +1,124 @@
+
+use trytes::*;
+use alloc::Vec;
+use curl::*;
+use cpucurl::*;
+use core::cmp::min;
+
+trait Offset {
+    fn offset(&mut self);
```
Comment (th0br0) [inline curl-cpu/src/search.rs]: Could you document this?

File: curl-cpu/src/search.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+impl<'a> Offset for &'a mut [BCTrit] {
+    fn offset(&mut self) {
+        self[0].0 = 0b1101101101101101101101101101101101101101101101101101101101101101;
```
Comment (th0br0) [inline curl-cpu/src/search.rs]: We probably want to explain how we end up at these magic numbers

File: curl-cpu/src/search.rs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        self[1].0 = 0b1111000111111000111111000111111000111111000111111000111111000111;
+        self[1].1 = 0b1000111111000111111000111111000111111000111111000111111000111111;
+        self[2].0 = 0b0111111111111111111000000000111111111111111111000000000111111111;
+        self[2].1 = 0b1111111111000000000111111111111111111000000000111111111111111111;
+        self[3].0 = 0b1111111111000000000000000000000000000111111111111111111111111111;
+        self[3].1 = 0b0000000000111111111111111111111111111111111111111111111111111111;
+    }
+}
+
+pub fn prepare_search(input: &[BCTrit]) -> Vec<BCTrit> {
```
Comment (th0br0) [inline curl-cpu/src/search.rs]: documentation

File: curl-cpu/src/search.rs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+#[cfg(feature = "parallel")]
+mod cpu_search {
+    use super::*;
+    use tmath::*;
+    use std::thread;
+    use std::sync::mpsc::channel;
+    use std::sync::atomic::{AtomicBool, Ordering};
+    use std::sync::Arc;
+    use core::marker::*;
+    //use alloc::Vec;
```
Comment (th0br0) [inline curl-cpu/src/search.rs]: unused

File: curl-cpu/src/search.rs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        }
+
+        let mux = TrinaryDemultiplexer::new(curl.squeeze(size).as_slice());
+
+        Some(mux[index.unwrap()].clone())
+    }
+}
+
+#[cfg(feature = "parallel")]
+mod cpu_search {
```
Comment (th0br0) [inline curl-cpu/src/search.rs:38]: my idea was to move this code to, e.g. `src/par/search.rs`  - just do a typedef somewhere in `lib.rs` e.g. at the level of the `pub use` at the end of this file?
I don't think that this is the only part of this crate that we want to parallelise?

File: tmath/src/increment.rs [hunk 6]
```diff
@@ -2,33 +2,35 @@ use trytes::*;
 
 /// Trait for incrementing trinary arrays
 pub trait TrinaryIncr {
-    fn incr(&mut self);
+    fn incr(&mut self) -> usize;
```
Comment (th0br0) [inline tmath/src/increment.rs]: document the return value?

File: curl/src/tests.rs [hunk 7]
```diff
@@ -146,6 +146,81 @@ mod inner {
...
+                                   BA9DFRQRLTLUJENKESDGTZRGRSLTNYTITXRXRGVLWBTEWPJXZYLGHLQBAVYV\
+                                   OSABIVTQYQM9FIQKCBRRUEMVVTMERLWOK"
+            .chars()
+            .collect();
+        let weight = 9u8;
+        let trans_copy: Trinary = trans.clone();
+        let nonce: Trinary = A::search(trans, weight).expect("Some PoW Failure.");
+        assert_eq!(nonce.trits().len(), 243);
+
+        let final_t: Trinary = trans_copy.trits()[..(trans_copy.len_trits() - HASH_LENGTH)]
```
Comment (th0br0) [inline curl/src/tests.rs]: As the hashing logic in here is rather fundamental, maybe we should use that to a `trait TrinaryHash` and implement for all `IntoTrinary`?

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: curl-cpu/src/search.rs (hunk 2 above)
Comment (paulhandy) [reply curl-cpu/src/search.rs]: I think it should really just be generated automagically for different word sizes.

**Comment by th0br0** [comment]
Sounds good. Could you file an issue? I'm happy with leaving it as is, just
don't want to lose track.

File: curl-cpu/src/search.rs (hunk 5 above)
Comment (paulhandy) [reply curl-cpu/src/search.rs:38]: This is the only part of the crate that we need to parallelize.

File: curl-cpu/src/search.rs (hunk 2 above)
Comment (th0br0) [reply curl-cpu/src/search.rs]: filed #70

File: curl-cpu/src/search.rs [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    use std::sync::Arc;
+    use core::marker::*;
+    pub fn search_cpu<F>(input: &[BCTrit], length: usize, group: usize, check: F) -> Option<Trinary>
+    where
+        F: Fn(&[BCTrit]) -> Option<usize> + 'static + Send + Sync,
+    {
+        let running = Arc::new(AtomicBool::new(true));
+        let check_arc = Arc::new(check);
+        let (tx, rx) = channel();
+        // TODO: replace with calculated number of threads
```
Comment (th0br0) [inline curl-cpu/src/search.rs]: Filed #71 could you remove the comment?

File: curl-cpu/src/search.rs [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    use std::thread;
+    use std::sync::mpsc::channel;
+    use std::sync::atomic::{AtomicBool, Ordering};
+    use std::sync::Arc;
+    use core::marker::*;
+    pub fn search_cpu<F>(input: &[BCTrit], length: usize, group: usize, check: F) -> Option<Trinary>
+    where
+        F: Fn(&[BCTrit]) -> Option<usize> + 'static + Send + Sync,
+    {
+        let running = Arc::new(AtomicBool::new(true));
```
Comment (th0br0) [inline curl-cpu/src/search.rs]: Does this need to be wrapped up in an `Arc`? Acces

**Review by th0br0 — APPROVED** [review]
Could you fix the nits and merge then?

**Comment by paulhandy** [comment]
merged

**Pull request closed without merging**